### DISCUSSION TITLE
Questions about decoration capabilities

### DISCUSSION THREAD

**Discussion by lukesmurray** [ueberdosis/tiptap#2263] [General Questions] [answered]
- Is it possible to use decorations to render custom (inline) react elements. For example, replacing all instances of `@username` with a custom element.
- Can you trigger decorations based on external data? For example, I want to highlight all words in a list, but that list comes from a server. Can I "redecorate" when the list changes even if the user hasn't changed the document?
- Can the decoration function be asynchronous? For example, calling a service worker with the current document contents, and rendering the results when the service worker returns?

**Comment by hanspagel** [comment]
1. Nope.
2. Yes.
3. Not sure about this, but sounds like you’d run into trouble.

Use that as a starting point: https://tiptap.dev/experiments/linter

**Comment by philippkuehn** [comment, marked as answer]
you can only render widget decorations as react components because it’s the only widget type that has a destroy method.

https://prosemirror.net/docs/ref/#view.Decoration^widget
